packages/contracts/test/utils/test_with_reference.ts

```diff
@@ -0,0 +1,110 @@
+    referenceFunc: (p0: P0, p1: P1, p2: P2, p3: P3, p4: P4) => Promise<R>,
+    testFunc: (p0: P0, p1: P1, p2: P2, p3: P3, p4: P4) => Promise<R>,
+    values: [P0, P1, P2, P3, P4],
+): Promise<void>;
+
+/**
+ * Tests the behavior of a test function by comparing it to the expected
+ * behavior (defined by a reference function).
+ *
+ * First the reference function will be called to obtain an "expected result",
+ * or if the reference function throws/rejects, an "expected error". Next, the
+ * test function will be called to obtain an "actual result", or if the test
+ * function throws/rejects, an "actual error". The test passes if at least one
+ * of the following conditions is met:
+ *
+ * 1) Neither the reference function or the test function throw and the
+ * "expected result" equals the "actual result".
+ *
+ * 2) Both the reference function and the test function throw and the "actual
+ * error" message *contains* the "expected error" message.
+ *
+ * @param referenceFuncAsync a reference function implemented in pure
+ * JavaScript/TypeScript which accepts N arguments and returns the "expected
+ * result" or throws/rejects with the "expected error".
+ * @param testFuncAsync a test function which, e.g., makes a call or sends a
+ * transaction to a contract. It accepts the same N arguments returns the
+ * "actual result" or throws/rejects with the "actual error".
+ * @param values an array of N values, where each value corresponds in-order to
+ * an argument to both the test function and the reference function.
+ * @return A Promise that resolves if the test passes and rejects if the test
+ * fails, according to the rules described above.
+ */
+export async function testWithReferenceFuncAsync(
+    referenceFunc: (...args: any[]) => Promise<any>,
+    testFuncAsync: (...args: any[]) => Promise<any>,
+    values: any[],
+): Promise<void> {
+    const paramNames = _getParameterNames(referenceFunc);
+    const testCaseString = JSON.stringify(_.zipObject(paramNames, values));
+    let expectedResult: any;
+    let expectedErr: string | undefined;
+    try {
+        expectedResult = await referenceFunc(...values);
+    } catch (e) {
+        expectedErr = e.message;
+    }
+    let actualResult: any | undefined;
+    try {
+        actualResult = await testFuncAsync(...values);
+        if (!_.isUndefined(expectedErr)) {
+            throw new Error(
+                `Expected error containing ${expectedErr} but got no error\n\tTest case: ${testCaseString}`,
+            );
+        }
+    } catch (e) {
+        if (_.isUndefined(expectedErr)) {
+            throw new Error(`${e.message}\n\tTest case: ${testCaseString}`);
+        } else {
+            expect(e.message).to.contain(expectedErr, `${e.message}\n\tTest case: ${testCaseString}`);
+        }
+    }
+    if (!_.isUndefined(actualResult) && !_.isUndefined(expectedResult)) {
+        expect(JSON.stringify(actualResult)).to.equal(
+            JSON.stringify(expectedResult),
```

Comment:
`JSON.stringify` is not a safe comparison. Order of properties can differ even though the objects are the same. Why not `expect(actualResult).to.deep.equal(expectedResult)`?

Reply:
`expect(actualResult).to.deep.equal(expectedResult)` results in output that is a lot harder to read when a test fails, especially for the `BigNumber` type, which is frequently used.

```
 |       AssertionError: Test case: {"numerator":"0","denominator":"1","target":"0"}: expected { Object (s, e, ...) } to deeply equal { Object (s, e, ...) }
 |       + expected - actual
 |
 |        {
 |          "c": [
 |       -    0
 |       +    10
 |          ]
 |       -  "e": 0
 |       +  "e": 1
 |          "s": 1
 |        }
 |
```

That said, `testWithReferenceFuncAsync` is supposed to work for any type, so it's important to use a reliable method of comparison. I went ahead and made this change.

If anyone else knows of a way to have both reliable comparison and readable output (particularly for `BigNumber` types) I'd be happy to hear suggestions.

Reply:
Seems like something we should patch in Chai.